scheds/rust/scx_mitosis/src/bpf/mitosis.bpf.c

```diff
@@ -0,0 +1,932 @@
+		return -ENOENT;
+	}
+
+	// Just initialize the cgroup lock
+	if (!lookup_cgrp_lock(cgrp))
+		return -ENOENT;
+
+	// Initialize the rootcg to cell 0
+	if (!cgrp->level) {
+		cgc->cell = 0;
+		return 0;
+	}
+
+	struct cgroup *parent_cg;
+	if (!(parent_cg = lookup_cgrp_ancestor(cgrp, 1)))
+		return -ENOENT;
+
+	struct cgrp_ctx *parent_cgc;
+	if (!(parent_cgc = lookup_cgrp_ctx(parent_cg))) {
+		bpf_cgroup_release(parent_cg);
+		return -ENOENT;
+	}
+
+	bpf_cgroup_release(parent_cg);
+	// Otherwise initialize to parent's cell
+	cgc->cell = parent_cgc->cell;
+	return 0;
+}
+
+void BPF_STRUCT_OPS(mitosis_cgroup_move, struct task_struct *p,
+		    struct cgroup *from, struct cgroup *to)
+{
+	struct task_ctx *tctx;
+
+	if (!(tctx = lookup_task_ctx(p)))
+		return;
+
+	if (p->scx.flags & SCX_TASK_QUEUED) {
+		quiescent(p, from);
+		runnable(p, tctx, to);
+	}
+
+	update_task_cell(p, tctx, to);
+}
+
+void BPF_STRUCT_OPS(mitosis_set_cpumask, struct task_struct *p,
+		    const struct cpumask *cpumask)
+{
+	struct task_ctx *tctx;
+
+	if (!(tctx = lookup_task_ctx(p)))
+		return;
+
+	if (!all_cpumask) {
+		scx_bpf_error("NULL all_cpumask");
+		return;
+	}
+
+	tctx->all_cpus_allowed = bpf_cpumask_subset(
+		(const struct cpumask *)all_cpumask, cpumask);
+}
+
+s32 BPF_STRUCT_OPS(mitosis_init_task, struct task_struct *p,
+		   struct scx_init_task_args *args)
+{
+	struct task_ctx *tctx;
+	int ret;
+
+	tctx = bpf_task_storage_get(&task_ctxs, p, 0,
+				    BPF_LOCAL_STORAGE_GET_F_CREATE);
+	if (!tctx) {
+		scx_bpf_error("task_ctx allocation failure");
+		return -ENOMEM;
+	}
+
+	if (all_cpumask)
+		tctx->all_cpus_allowed = bpf_cpumask_subset(
+			(const struct cpumask *)all_cpumask, p->cpus_ptr);
+	else {
+		scx_bpf_error("missing all_cpumask");
+		return -EINVAL;
+	}
+
+	if ((ret = update_task_cell(p, tctx, args->cgroup))) {
+		return ret;
+	}
+
+	return 0;
+}
+
+s32 BPF_STRUCT_OPS_SLEEPABLE(mitosis_init)
+{
+	struct bpf_cpumask *cpumask;
+	u32 i;
+	s32 ret;
+
+	__COMPAT_scx_bpf_switch_all();
+
+	cpumask = bpf_cpumask_create();
+	if (!cpumask)
+		return -ENOMEM;
+
+	bpf_for(i, 0, nr_possible_cpus)
+	{
+		const volatile u8 *u8_ptr;
+
+		if ((u8_ptr = MEMBER_VPTR(all_cpus, [i / 8]))) {
+			if (*u8_ptr & (1 << (i % 8))) {
+				bpf_cpumask_set_cpu(i, cpumask);
+			}
+		} else {
+			return -EINVAL;
+		}
+	}
+
+	cpumask = bpf_kptr_xchg(&all_cpumask, cpumask);
+	if (cpumask)
+		bpf_cpumask_release(cpumask);
+
+	bpf_for(i, 0, MAX_CELLS)
+	{
+		struct cell_cpumask_wrapper *cpumaskw;
+		struct cell *cell = &cells[i];
+
+		ret = scx_bpf_create_dsq(i, -1);
+		if (ret < 0)
+			return ret;
+		cell->dsq = i;
+
+		if (!(cpumaskw = bpf_map_lookup_elem(&cell_cpumasks, &i)))
+			return -ENOENT;
+
+		cpumask = bpf_cpumask_create();
+		if (!cpumask)
+			return -ENOMEM;
+
+		/*
+		 * Start with all full cpumask for all cells. It only matters for cell 0
+		 * to start with, all others will get reconfigured by userspace before
+		 * being used.
+		 */
+		bpf_cpumask_setall(cpumask);
```

Comment:
Hmm, do you want `bpf_cpumask_copy(cpumask, all_cpumask);`?

Reply:
Good catch, will fix.